kong/config_service.go

```diff
@@ -0,0 +1,49 @@
+package kong
+
+import (
+	"context"
+	"fmt"
+	"io"
+)
+
+// AbstractConfigService handles Config in Kong.
+type AbstractConfigService interface {
+	ReloadDeclarativeConfig(ctx context.Context, config any, checkHash int) error
```

Comment:
Would be great if we could somehow narrow down the type of `config`, ideally making it a strongly typed configuration 🤔 Allowing to pass any type does make this method very easy to break. 

If that's not possible to make it a strong type, let's maybe at least leave a comment describing what we're actually expecting it to be (e.g. "declarative config request body").

Reply:
I thought about that but there are a couple of points that make this "challenging":

- you can't add type parameters to methods to something like this:

  ```
  type configConstraint interface {
	  []byte | string | Configuration | io.Reader
  }
  ```

  is not allowed. You either specify a union of structs (any of those) and add an interface **which also should be fulfilled** or use 1 of those constraints. Can't make it an alternative. https://github.com/golang/go/issues/49054#issuecomment-946337356

- [`NewRequestRaw()`](https://github.com/Kong/go-kong/blob/d5120140735c43bb5f14f6cd2cb0cbe9f8d76d77/kong/request.go#L14) accepts `interface{}` and is widely used (and exported) so we can't just will nilly change this (or can we? we're pre v1 :) ). This is not a problem per se but would be a nice improvement to have the whole call stack guarded by type constraints.

---

PTAL at 535869c, I've added a second method which allows sending the config as `Configuration` whereas the original was adopted to accept `io.Reader`. I believe that's the best of 2 worlds for now.

Reply:
Couldn't we leave just the strongly typed method? Is a reason for leaving the loosely typed `ReloadDeclarativeConfig` that we (in KIC) don't use the strong type when sending the config?

Reply:
It's a bit more nuanced 🤯 

In KIC we use [`*file.Content`](https://github.com/Kong/kubernetes-ingress-controller/blob/aac834bd06a89d488f61d927fb5541413e65499b/internal/dataplane/sendconfig/sendconfig.go#L125) from [deck](https://github.com/Kong/deck/blob/be0eaa0835c8ca90fb5944481defe6ac534a41be/file/types.go#L660). We can't use it in `go-kong` because that would create a circular dependency ([deck already imports go-kong](https://github.com/Kong/deck/blob/be0eaa0835c8ca90fb5944481defe6ac534a41be/go.mod#L22)).

Ideally this would be the type we use there but 🤷 I have no idea (other than using something else like `map[string]interface{}` :( or copy pasting) how to approach this properly.

Hence why I left the `io.Reader` as an option to pass serialized version of `*file.Content`, `Configuration`, `map[string]interface{}` or anything else if the user's  brave enough.

Reply:
Hmm, got it. So maybe just to make the distinction between these two more favoring the strongly typed one, we could rename them as follows:
- `ReloadDeclarativeConfig` -> `ReloadRawDeclarativeConfig` (the Raw part should be discouraging usage for new users)
- `ReloadDeclarativeConfigWithSchema` -> `ReloadDeclarativeConfig` (the right ™️ one)
WDYT?

Reply:
As we've discovered `kong.Configuration` is not really a type for the whole kong configuration, but a single plugin, I propose to keep only the `ReloadRawDeclarativeConfig` accepting an `io.Reader` for now.

Reply:
Done.